### PR TITLE
Enable generalized LayerNorm on CUDA

### PR DESCRIPTION
**Pull request by BBC-Esq**
Pertains to https://github.com/OpenNMT/CTranslate2/issues/1954
> Consulted this in part but apparently the three tiered approach isn't great for whisper except if layernorm is the last layer or something:  https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/cuda/layer_norm.cuh
> And an article link in plain english: https://oneflow2020.medium.com/how-to-implement-an-efficient-layernorm-cuda-kernel-oneflow-performance-optimization-731e91a285b8

---

In the Whisper alignment path, `attention_probs` is already fully computed on the GPU before it is explicitly transferred back to the CPU. This can be verified by following the execution flow:

1. **`attention_weights` is produced by the decoder on the model device**
   When the model runs on CUDA, the decoder outputs `attention_weights` as a CUDA tensor.

2. **SoftMax runs on CUDA**

   ```cpp
   ops::SoftMax()(attention_weights);
   ```

   `SoftMax` has a CUDA implementation and dispatches to a GPU kernel when the input is on CUDA.

3. **Padding removal runs on CUDA**

   ```cpp
   remove_padding(attention_weights, ...);
   ```

   This uses `Split`, which has a CUDA implementation and does not force a device change.

4. **Batch selection runs on CUDA**

   ```cpp
   ops::Gather()(attention_weights, batch_id, attention_probs);
   ```

   `Gather` defaults to `axis=0, batch_dims=0`, which is explicitly supported by the CUDA implementation. This keeps `attention_probs` on CUDA.

At this point, **`attention_probs` is still a CUDA tensor**.

However, `compute_alignments` then unconditionally does:

```cpp
attention_probs.move_to(Device::CPU, DataType::FLOAT32);
```

This forces a GPU → CPU transfer regardless of whether the preceding operations could run on GPU.

---

### Impact of enabling GPU LayerNorm

With generalized LayerNorm now implemented on CUDA, the call:

```cpp
ops::LayerNorm(-2, 0)(attention_probs);
```

can safely execute on GPU **before** the CPU transfer. This removes an unnecessary CPU fallback for LayerNorm and ensures that:

* LayerNorm computation itself no longer runs on the CPU
* The only remaining GPU → CPU transfer is for subsequent operations that are still CPU-only (e.g., MedianFilter, DTW).

This does **not** make the entire alignment path GPU-only, but it:

* Eliminates a hard GPU limitation in LayerNorm
* Preserves correctness
* Enables incremental performance improvements without changing higher-level behavior

---

### What remains to fully eliminate the CPU transfer

Even with generalized LayerNorm now supported on CUDA, compute_alignments still contains operations that require CPU execution, which necessitates the remaining GPU → CPU transfer. In particular, the median filtering step (ops::MedianFilter) and the dynamic time warping alignment (negative_dtw) are currently CPU-only. While the subsequent reduction (ops::Mean) already has a CUDA implementation, it cannot be reached today because the alignment pipeline explicitly moves tensors to CPU before these stages. This change therefore focuses on removing an unnecessary GPU limitation in LayerNorm, while preserving the existing CPU-based alignment logic. Fully eliminating the CPU transfer would require additional GPU implementations of the remaining CPU-only stages or a restructuring of the alignment computation.

### CODE REVIEW

**Comment by BBC-Esq** [comment]
For the other things running on CPU-only here are some decent links:

## DTW

https://github.com/qianlikzf/GPUDTW

https://github.com/Maghoumi/pytorch-softdtw-cuda

https://www.researchgate.net/publication/343707254_cuDTW_Ultra-Fast_Dynamic_Time_Warping_on_CUDA-Enabled_GPUs

## Median Filtering

https://docs.nvidia.com/vpi/algo_median_filter.html

https://github.com/jonaylor89/Median-Filter-CUDA

https://www.researchgate.net/publication/327855146_Efficient_GPU-based_implementation_of_the_median_filter_based_on_a_multi-pixel-per-thread_framework

https://jmcms.s3.amazonaws.com/wp-content/uploads/2022/09/24093435/52-jmcms-2008112-Investigating-the-Median-Filter-Operation-Sheeraz-3-8-7-2020.pdf

**Comment by jordimas** [comment]
CTranslate2 is a low-level and complex codebase.

As a maintainer, I am not in favor of AI-generated code for several reasons:

* It shifts the burden to me to double-check everything, including correctness, testing, performance, and fully understanding the code. It takes some minutes to use AI, hours to do the rest of the process.
* It discourages contributions from people with real expertise, as they may think that AI-generated submissions actually fix an issue.
* I expect contributors to have a deep understanding of what the code does and why it is written that way.

Going forward, I will:

* Refuse pull requests that are clearly written by AI.
* Ignore suggestions that are based purely on what an AI says, without proper understanding.

I know these contributions are made in good faith. I am happy to continue working with you on adding examples, improving documentation, or contributing to Hugging Face-related integrations, but these should be areas that match the contributor’s level of expertise.

I will do PR to add this to CONTRIBUTING.md then the guidelines are clear to everybody.

Note: I take code maintainability and understanding as key priority for this project, and you will see how in many PR I expect people to discuss what's the goal of every single line (see sample https://github.com/OpenNMT/CTranslate2/pull/1903).

**Comment by BBC-Esq** [comment]
> CTranslate2 is a low-level and complex codebase.

Agreed!

> As a maintainer, I am not in favor of AI-generated code for several reasons:
>     It shifts the burden to me to double-check everything, including correctness, testing, performance, and fully understanding the code. It takes some minutes to use AI, hours to do the rest of the process.

I can test as well before submitting a PR if that would obviate this issue?  Despite what @Purfview believes, I don't simply copy and paste whatever AI generates and I do feel like I understand all of my pull requests.  I understand my code!

> It discourages contributions from people with real expertise, as they may think that AI-generated submissions actually fix an issue.

I agree that simply pasting "AI-generated" submissions isn't appropriate.  However, "AI-assisted" solutions actual can and do help programmers (key word "help") long as the person doesn't abdicate their duty of reviewing and understanding.  It can greatly accelerate writing code...

See here an example from @jncraton: https://github.com/OpenNMT/CTranslate2/issues/1946#issuecomment-3678187280

image

This PR, and my others, are based on what I view as a full understanding of what I'm submitting, again, despite what @Purfview would like to portray me as. I'm doing the same workflow that jncranton does, essentially, among a vast majority of other programmers, using AI to "assist" where appropriate.  I do understand the PRs that I submit.

>    I expect contributors to have a deep understanding of what the code does and why it is written that way.

See my comment above.  At this point I would characterize my understanding of programming as "deep."  See here for one example: https://github.com/OpenNMT/CTranslate2/issues/1950#issuecomment-3676932290  If I made a mistake in this PR perhaps we could discuss it would somehow prove my understanding to you?
> 
> Going forward, I will:
> 
>    Refuse pull requests that are clearly written by AI.

To clarify, I use AI to "assist" or "recommend" changes to actual code, but I always understand the code I submit.  I have continued to do this as I have taught myself programming over the last 2-3 years.  Also, I will sometimes (not always) use AI to summarize the nature of a PR to save time in writing the summary myself, but that's separate and distinct from whether the actual code changes are correct and I can refrain from that in the future if need be...

>    Ignore suggestions that are based purely on what an AI says, without proper understanding.

100% agree with this.  See my immediately preceding comments.
> 
> I know these contributions are made in good faith. I am happy to continue working with you on adding examples, improving documentation, or contributing to Hugging Face-related integrations, but these should be areas that match the contributor’s level of expertise.

I would submit that I have the requisite level of expertise and ask that my level of expertise not be pre-judged simply because I've taught myself programming as a hobby and don't have a college degree in it.  I understand the code I submit.  If a formal education in programming is required that's a different story.  I've taught myself programming the last 2-3 years at least and you're going to have varying levels of "expertise" among contributors, and I would like to feel that my level of expertise is sufficient by now...

> I will do PR to add this to CONTRIBUTING.md then the guidelines are clear to everybody.
> 
> Note: I take code maintainability and understanding as key priority for this project, and you will see how in many PR I expect people to discuss what's the goal of every single line (see sample #1903).

Understood.  I'm always open to discussing why I wrote a PR a certain way or any of the code in it.

**Comment by Purfview** [comment]
>This PR, and my others, are based on what I view as a full understanding of what I'm submitting

But your PRs says otherwise, you don't understand your commits. It's not "Purfview believes", it's a fact.

You should work on managing your anger and frustration, and learn to accept "no".

**Comment by BBC-Esq** [comment]
Again, @Purfview your comment of: "anger and frustration, and learn to accept 'no'" is unprofessional.  What is your problem with me? I stand by what I wrote to @jordimas and stand ready to verify it.  That said, I have no desire to argue further with you and having nothing to prove to you.  Leave me alone.  Thanks.

**Comment by Purfview** [comment]
>Leave me alone.

Wait, but it's YOU who pinged me in your comments...

**Comment by BBC-Esq** [comment]
You're commenting on me in my faster-whisper PR and have followed me here to harass me and my work further.  You've been doing this behavior for over a year now.  Just stop.  Calling me a liar as if I'm lying to @jordimas about understanding something...Again, just back off please.

**Comment by Purfview** [comment]
> followed me

**YOU** pinged me here.

> harass me

It's YOU harassing repo owners. God knows on how many you are blocked...

**Comment by jordimas** [comment]
I think that I have been clear on my criteria now and moving forward. Closing the PR.

**Comment by BBC-Esq** [comment]
@jordimas I am not clear.  What was improper with this PR specifically so I understand moving forward?

**Comment by a2d8a4v** [comment]
@BBC-Esq I think jordimas has declared clearly the concern about generating most of code with AI tools without deep understanding. Basing on my experience at least, these AI tools do not really understand what we really need, we still needs to take time to check its correctness and logical issues.
And when encountering such big changes like this PR, we should take it seriously. And you have said that you did not evaluate your code works or not, It's hard for people to believe and hard to accept this PR.

**Comment by BBC-Esq** [comment]
@a2d8a4v I did evaluate my code and I even offered to demonstrate my understanding.  And as far as I understand it, there's nothing wrong with this PR, although I'm open to being wrong.  I am awaiting a response as to why this PR was not even reviewed.  For example, does it not actually "fix the issue," which was a concern  Was it because I don't have a "deep understanding of what the code does" or because it's "based purely on what AI said, without proper understanding" as @jordimas is concerned about?  Again, I'm available to answer any questions about it.

**Comment by a2d8a4v** [comment]
@BBC-Esq Hi, I think at least I can response some philosophy and manners about coding related to @jordimas 's concerns.
"deep understanding of what the code does" matters, cause reviewers need to discuss with the authors who pulled the PRs.

If reviewers can not check the code, for instance the correctness and logic issues, the quality of codebase drops, and some invisible bugs would be triggered in the future, and may be hard to find out (especially CTranslate2 using compiling language, which will take more time). It will waste other people's energy and time to just target and fix those bugs. It may take 10x~100x of time to drive those issues basing on the complexity of the changes, etc.

By the way, while coding, it's better to give some comments for your code sections so that people can easily understand.

In conclude, I reckon that the PR's author should be responsible for what they did. Using AI toolkits is great, yes it's convenient, but it's just a tool. For a sustainable project, the intervention of people is vital.

**Comment by jordimas** [comment]
> I will do PR to add this to CONTRIBUTING.md then the guidelines are clear to everybody.

Here we go: https://github.com/OpenNMT/CTranslate2/pull/1961

**Comment by BBC-Esq** [comment]
@a2d8a4v I appreciate your message and will try to add comments in the future.  I'd still like to know from @jordimas what about my PR warranted not even a review.  Was it incorrect or something?  As far as I could tell it works, which is the most important thing.

**Comment by hydra3333** [comment]
As a retired programmer (ie has a bit of experience) ... usually there are documented test plans dealing with a variety of normal use-cases and as many edge cases that you can think of and also cases which other interested people may contribute, with documented results arising from each test to prove that the change works as expected in every case. Perhaps you did do such a thing - however with 2-3 years self-taught, I am unsure if you appreciate the necessity for a stringent verification approach to programming.  Just a thought.
